index.js

```diff
@@ -580,6 +593,150 @@ module.exports = {
 
     self.rewriteUrl = url => {
       return url;
-    }
+    };
+
+    self.getPagesTree = async (req) => {
+      self.maps = {};
+      const excludedTypes = [
+        'workflow-document',
+        ...self.options.excludeTypes || []
+      ]
+
+      const pages = await getPages();
+      const pagesWithPieces = await getPieces(pages);
+      const pagesTree = buildPagesTree(pagesWithPieces);
+
+      return pagesTree;
+
+      async function getPages () {
```

Comment:
The logic to build the tree etc. can all be replaced with:

```
await self.apos.pages.find(req, { level: 0 }, { projection... }).children({ depth: 1000, projection: { ... }).toObject();
```

Which will return the home page with a `_children` array and its subpages will have `_children` arrays if appropriate and so on.

Reply:
Indeed I can use this method, it would add extra work because we need then to parse recursively the tree in order to add the pieces in the right page, this part is not big deal.
My concern is about `excludeTypes`, I must do the same recursive check here, which is still no big deal too. But what should I do if a page is excluded but has children that are not? Should we remove them too? In the current code they are kept as normal page with no parent.

Reply:
Interesting, I take your point about locating the right piece page in the tree, but walking the tree of pages to find those is still probably less work than building it correctly from scratch.

You can pass any query builder as a property of the options object given to `children`, which includes using `and: { ... }` to pass in some mongodb criteria based on excludeTypes.